### ISSUE TITLE
[Hacktoberfest 2020] Update stubs for new `model-property` type

### ISSUE THREAD

**Issue by canvural** [nunomaduro/larastan#676]
## The Problem
With the new `model-property` type and `ModelPropertyRule` Larastan now has the ability to check the existence of model columns that are given to a method.

Adding the `model-property` type to the Laravel's core methods are done with the help of [PHPstan stubs](https://phpstan.org/user-guide/stub-files). Some [examples](https://github.com/nunomaduro/larastan/blob/master/stubs/EloquentBuilder.stub#L14) can be seen [here](https://github.com/nunomaduro/larastan/blob/master/stubs/Model.stub#L13). Currently, not all the Laravel methods that are doing something with the model properties are updated in stubs. And we need help with updating this.

## Contributing

You can contribute to this issue by updating the stubs and tests. Below you can read the detailed instructions for how to that. In case you have any questions you can ask them in this issue, or you can reach out to me at [Twitter](https://twitter.com/can__vural)

### Preparation

* Create a [GitHub user account](https://www.wikihow.com/Create-an-Account-on-GitHub) (if you haven't got one already).
* [Fork this repository on GitHub](https://help.github.com/en/articles/fork-a-repo).
* Make sure you have both [git](https://git-scm.com/downloads) as well as [Composer](https://getcomposer.org/) installed on your machine.
* [Clone your fork to your local machine](https://help.github.com/en/articles/cloning-a-repository) using your preferred git client.
* Run `composer install` from the command line in the root directory of your local clone of this repo.

#### Helpful guides
* https://www.digitalocean.com/community/tutorials/hacktoberfest-how-to-submit-your-first-pull-request-on-github
* https://www.digitalocean.com/community/tech_talks/level-up-as-an-open-source-contributor
* https://www.nexmo.com/blog/2020/10/06/opening-multiple-pull-requests-dr

### Updating stubs

[Here](https://github.com/nunomaduro/larastan/tree/master/stubs) is the directory containing all of the stubs Larastan uses. All of the stubs that need updating are already there, so you don't need to create new files. Just updating method definitions inside.

#### Steps
1. Find a Laravel method that does not exist in our stubs, or exists but the parameter type is not yet updated. For this example, we will use this method here: https://github.com/laravel/framework/blob/8.x/src/Illuminate/Database/Eloquent/Builder.php#L511

2. Then we need to copy the method definition into our stub. So in [this](https://github.com/nunomaduro/larastan/blob/master/stubs/EloquentBuilder.stub#L169) file, at the end we can add the following code:
```php
/**
 * Get a single column's value from the first result of a query.
 *
 * @param  model-property<TModelClass>  $column
 * @return mixed
 */
public function value($column);
```

Here we used generic type `model-property<TModelClass>` because we only want to allow the model properties from the current model. You can [see](https://github.com/nunomaduro/larastan/blob/master/stubs/EloquentBuilder.stub#L6) the `TModelClass` defined at the beginning of the class. _If the stub you are updating does not have any `@template` definition for the class you can just use `model-property`_

3. As a last step we need to update the tests to verify our changes are working as we expect.
The method we added to stubs was on Eloquent Builder class, so we need to add our test case to the [`model-property-builder`](https://github.com/nunomaduro/larastan/blob/master/tests/Rules/Data/model-property-builder.php) file. If the stu you added was in one of the relation classes you can use [`model-property-relation`](https://github.com/nunomaduro/larastan/blob/master/tests/Rules/Data/model-property-relation.php) file. If it was in the model class itself, you can use [`model-property-model`](https://github.com/nunomaduro/larastan/blob/master/tests/Rules/Data/model-property-model.php) file.

 Our test case can be as simple as the following:
```php
\App\User::query()->value('emaiil');
```

 We intentionally made a typo and we expect Larastan to catch it.

**Note:** Here you can also add test cases that shouldn't produce any error. If you want to test that aspect of the method too.

Now that we added a test case, we also need to add an assertion that the expected error is produced. For that we need to add a new line to the [`testModelPropertyRuleOnBuilder`](https://github.com/nunomaduro/larastan/blob/master/tests/Rules/ModelPropertyRuleTest.php#L11) method in [`ModelPropertyRuleTest` ](https://github.com/nunomaduro/larastan/blob/master/tests/Rules/ModelPropertyRuleTest.php)class.

```php
public function testModelPropertyRuleOnBuilder(): void
{
    $errors = $this->setConfigPath(__DIR__.DIRECTORY_SEPARATOR.'Data/modelPropertyConfig.neon')->findErrorsByLine(__DIR__.'/Data/model-property-builder.php');

    self::assertEquals([
        // Rest of the assertions
        35 => 'Property \'emaiil\' does not exist in App\\User model.', // Newly added line
    ], $errors);
}
```

Here the array key represents the line number of the error. And the value is the message we expect to be produced.

Now let's run the whole test suit to see we did not break anything else. You can do that by running the command `composer test` in your terminal. If everything goes well you should see all green:

![image](https://user-images.githubusercontent.com/1574232/95329958-14b70a00-08a8-11eb-8126-a9d177db1ea8.png)

Then you can commit your changes, push it to your fork of the repo then go onto Github and open a pull request!

That's it! You've succesfully created a PR on an open source project 🥳

(_PS: The example used in this issue is not currently in the stubs. So you can just follow the guide, copy the code and make an easy first contribution_ 😊)

**Comment by xitox97** [comment]
Hi @canvural, thank you for writing this awesome instruction. It is really helpful for beginners like me. 

I have a question. Does 1 PR can only contains 1 method or 1 PR can contains multiple method?

**Comment by szepeviktor** [comment]
Thank you @xitox97 for your intention.
You may send as many stubs as you want in one PR.

**Comment by canvural** [comment]
@xitox97 Yeah it is ok to add more than one method in PR :+1:

**Comment by robbplo** [comment]
I've added two stubs in this PR: https://github.com/nunomaduro/larastan/pull/679

I didn't link the issue so it wouldn't close if it gets merged, because there is more work to be done!
Let me know if there's some changes I still need to make

**Comment by xitox97** [comment]
Hi @canvural , I just fork this repo and run composer test on master branch. I got a fail test for my PHPUnit. I already run composer install. Is there anything that I'm doing wrongly?

![image](https://user-images.githubusercontent.com/15958161/95650395-0b32d980-0b16-11eb-9b8b-b0e5df083384.png)

I'm using windows 10.

**Comment by szepeviktor** [comment]
Hello @xitox97! Please open a proper issue.

**Comment by xitox97** [comment]
> Hello @xitox97! Please open a proper issue.

Ops sorry, I thought can just ask it here. I open the issue here #681

**Comment by romanstingler** [comment]
will this syntax be supported in the future?
```
/**
 * @property int id
 * @property string name
 * @property string email
 * @property string password
 * @property string timezone
 * @property boolean is_active
 * @property-read Collection services
 */
class User extends Authenticatable
```

**Comment by canvural** [comment]
> will this syntax be supported in the future?
> 
> ```
> /**
>  * @property int id
>  * @property string name
>  * @property string email
>  * @property string password
>  * @property string timezone
>  * @property boolean is_active
>  * @property-read Collection services
>  */
> class User extends Authenticatable
> ```

what do you mean exactly? Yes those are model properties and will be checked.

**Comment by romanstingler** [comment]
![2024-11-25-16:45:11](https://github.com/user-attachments/assets/169d2c32-8c95-4988-a2f0-d6c8b8d0486f)
![2024-11-25-16:46:08](https://github.com/user-attachments/assets/4f2e216d-777c-4660-aa7f-0f9f393a4739)

```phpstan: Access to an undefined property App\Models\InventoryAdjustment::$status.```

**Comment by canvural** [comment]
This is not related to `model-property` at all.

Your issue is that all your variable names are missing `$` at the beginning.

**Comment by xtrasmal** [comment]
@romanstingler 

> `phpstan: Access to an undefined property App\Models\InventoryAdjustment::$status.`

### TLDR: 
**Your screenshot says all. ** Showing the exact problem. 
- **PHPSTAN:** tells you exactly what is up.
-  **PHPSTORM:** underlines what PHPSTAN is saying.

That's 2 HINTS in one image!
 - that should have triggered you NOT to post a question here and debug locally. 

### **TIPS** 
> [!TIP]
> **Assume that the error is literally what is wrong**
> Whenever you see an error like `phpstan: Access to an undefined property $status` 
> _Debug the literal error presented in your face, until you can proof that the root cause is something else._ 

@romanstingler this is what both **PHPSTAN** and **PHPSTORM** are **literally** telling you:

```php
<?php
# wrong
/** @property string status */
class MyClass {}
# correct
/** @property string $status */
class MyClass {}
```

> [!IMPORTANT]
> **Assume your IDE underlines hints are trying to tell you something**
> PHPStorm is losing its mind(all the coloured lines). 
> It is screaming something needs attention.
> Find out why and don't give up until you find out.
> If you think it is a IDE bug, then proof it by writing a unit test with expectations. 
> You can clear the IDE's cache and restart. Might be a stale index.  **Else follow these IDE hints..**. 
> It's why you have an  IDE in the > first place.
